can't fmt scal with `\ud801`,which actually is  `?`

##  Description
orginal code in pekko-grpc
org.apache.pekko.grpc.scaladsl.headers.HeadersSpec

```scala
  "Status-Message.value()" should {
    "use percent-encoding" in {
      // test cases taken from https://github.com/grpc/grpc-java/blob/79e75bace40cea7e4be72e7dcd1f41c3ad6ee857/api/src/test/java/io/grpc/StatusTest.java#L65
      val inAndExpectedOut = Table(
        ("raw input", "expected encoded value"),
        ("my favorite character is i", "my favorite character is i"),
        ("my favorite character is \n", "my favorite character is %0A"),
        ("my favorite character is \u0000", "my favorite character is %00"),
        ("my favorite character is %", "my favorite character is %25"),
        ("my favorite character is 𐀁", "my favorite character is %F0%90%80%81"),
        // \ud801 is a high surrogate, a lone surrogate character is getting decoded as ? with UTF-8
        ("my favorite character is \ud801", "my favorite character is ?"),
        // \udc37 is a low surrogate, a lone surrogate character is getting decoded as ? with UTF-8
        ("my favorite character is \udc37", "my favorite character is ?"),
        // a pair of surrogate characters is fine
        ("my favorite character is " + 0xDBFF.toChar + 0xDFFF.toChar, "my favorite character is %F4%8F%BF%BF"))

      forAll(inAndExpectedOut) { (in, expected) =>
        new `Status-Message`(in).value() should equal(expected)
      }
    }
  }
```

when scalafmtAll

```shell

[error] org.scalafmt.sbt.ScalafmtSbtReporter$ScalafmtSbtError: scalafmt: /home/laglang/coding/pekko-grpc/runtime/src/test/scala/org/apache/pekko/grpc/scaladsl/headers/HeadersSpec.scala:39: error: invalid unicode surrogate pair
[error]         ("my favorite character is \ud801", "my favorite character is ?"),
[error]                                    ^ [/home/laglang/coding/pekko-grpc/runtime/src/test/scala/org/apache/pekko/grpc/scaladsl/headers/HeadersSpec.scala]
[error] (runtime / Test / scalafmt) org.scalafmt.sbt.ScalafmtSbtReporter$ScalafmtSbtError: scalafmt: /home/laglang/coding/pekko-grpc/runtime/src/test/scala/org/apache/pekko/grpc/scaladsl/headers/HeadersSpec.scala:39: error: invalid unicode surrogate pair
[error]         ("my favorite character is \ud801", "my favorite character is ?"),
[error]                                    ^ [/home/laglang/coding/pekko-grpc/runtime/src/test/scala/org/apache/pekko/grpc/scaladsl/headers/HeadersSpec.scala]
[error] Total time: 3 s, completed 2024年4月14日 下午2:38:01

```


## Question
pekko-grpc can fix it by use `${0xD801.toChar}`, I'm not sure if this is a bug in scalafmt, report it.

 `\ud801` actually is  `?`

--- a/scalameta/tokenizers/shared/src/main/scala/scala/meta/internal/tokenizers/CharArrayReader.scala
+++ b/scalameta/tokenizers/shared/src/main/scala/scala/meta/internal/tokenizers/CharArrayReader.scala
@@ -55,21 +55,14 @@ private[meta] case class CharArrayReader private (
     else {
       begCharOffset = endCharOffset
       val (hi, hiEnd) = readUnicodeChar(buf, endCharOffset)
-      val isHiSurrogate = Character.isHighSurrogate(hi)
-      if (isHiSurrogate && hiEnd >= buf.length)
-        readerError("invalid unicode surrogate pair", at = begCharOffset)
-      else if (!isHiSurrogate || ch == '\'' && buf(hiEnd) == '\'' && !wasMultiChar) {
-        ch = hi
-        endCharOffset = hiEnd
-      } else {
+      endCharOffset = hiEnd
+      if (hiEnd < buf.length && Character.isHighSurrogate(hi)) {
         val (lo, loEnd) = readUnicodeChar(buf, hiEnd)
-        if (!Character.isLowSurrogate(lo))
-          readerError("invalid unicode surrogate pair", at = begCharOffset)
-        else {
+        if (Character.isLowSurrogate(lo)) {
           ch = Character.toCodePoint(hi, lo)
           endCharOffset = loEnd
-        }
-      }
+        } else ch = hi
+      } else ch = hi
     }
 
   def nextNonWhitespace = {

--- a/tests/shared/src/test/scala/scala/meta/tests/tokenizers/BaseTokenizerSuite.scala
+++ b/tests/shared/src/test/scala/scala/meta/tests/tokenizers/BaseTokenizerSuite.scala
@@ -24,11 +24,17 @@ abstract class BaseTokenizerSuite extends TreeSuiteBase {
 
   def assertTokenizedAsStructureLines(code: String, expected: String, dialect: Dialect = Scala211)(
       implicit loc: Location
-  ): Unit = assertNoDiff(tokenize(code, dialect).map(_.structure).mkString("\n"), expected)
+  ): Unit = assertTokensAsStructureLines(tokenize(code, dialect), expected)
+
+  def assertTokensAsStructureLines(tokens: Tokens, expected: String)(implicit loc: Location): Unit =
+    assertNoDiff(tokens.map(_.structure).mkString("\n"), expected)
 
   def assertTokenizedAsSyntax(code: String, expected: String, dialect: Dialect = Scala211)(implicit
       loc: Location
-  ): Unit = assertNoDiff(tokenize(code, dialect).map(_.syntax).mkString, expected)
+  ): Unit = assertTokensAsSyntax(tokenize(code, dialect), expected)
+
+  def assertTokensAsSyntax(tokens: Tokens, expected: String)(implicit loc: Location): Unit =
+    assertNoDiff(tokens.map(_.syntax).mkString, expected)
 
   implicit class ImplicitString(value: String) {
     def tq(repl: String): String = value.replace(repl, "\"\"\"")

--- a/tests/shared/src/test/scala/scala/meta/tests/tokenizers/TokenizerSuite.scala
+++ b/tests/shared/src/test/scala/scala/meta/tests/tokenizers/TokenizerSuite.scala
@@ -1642,4 +1642,56 @@ class TokenizerSuite extends BaseTokenizerSuite {
     }
   }
 
+  Seq(
+    // high-surrogate char codes
+    55297
+  ).foreach { ch =>
+    test(s"#3690 high-surrogate with char code $ch") {
+      def tokens = tokenize(
+        s"""|"${ch.toChar}"
+            |""".stripMargin,
+        dialects.Scala213
+      )
+      val exception = intercept[TokenizeException](tokens)
+      val err = "<input>:1: error: invalid unicode surrogate pair"
+      assertEquals(exception.getMessage.substring(0, err.length), err)
+    }
+  }
+
+  Seq(
+    // regular or low-surrogate or char codes
+    56375
+  ).foreach { ch =>
+    test(s"#3690 non-high-surrogate with char code $ch") {
+      tokenize(
+        s"""|"${ch.toChar}"
+            |""".stripMargin,
+        dialects.Scala213
+      ) match {
+        case Tokens(_: BOF, Constant.String(str), _: LF, _: EOF) =>
+          assertEquals(str.length, 1)
+          assertEquals(str, s"${ch.toChar}")
+        case tokens => fail(s"unexpected tokens: ${tokens.structure}")
+      }
+    }
+  }
+
+  Seq(
+    // full-surrogate char code pairs
+    (55297, 56375),
+    (0xdbff, 0xdfff)
+  ).foreach { case (hi, lo) =>
+    test(s"#3690 full-surrogate with char codes $hi and $lo") {
+      tokenize(
+        s"""|"${hi.toChar}${lo.toChar}"
+            |""".stripMargin,
+        dialects.Scala213
+      ) match {
+        case Tokens(_: BOF, Constant.String(str), _: LF, _: EOF) =>
+          assertEquals(str, s"${hi.toChar}${lo.toChar}")
+        case tokens => fail(s"unexpected tokens: ${tokens.structure}")
+      }
+    }
+  }
+
 }

--- a/tests/shared/src/test/scala/scala/meta/tests/tokenizers/TokenizerSuite.scala
+++ b/tests/shared/src/test/scala/scala/meta/tests/tokenizers/TokenizerSuite.scala
@@ -1643,26 +1643,11 @@ class TokenizerSuite extends BaseTokenizerSuite {
   }
 
   Seq(
-    // high-surrogate char codes
-    55297
-  ).foreach { ch =>
-    test(s"#3690 high-surrogate with char code $ch") {
-      def tokens = tokenize(
-        s"""|"${ch.toChar}"
-            |""".stripMargin,
-        dialects.Scala213
-      )
-      val exception = intercept[TokenizeException](tokens)
-      val err = "<input>:1: error: invalid unicode surrogate pair"
-      assertEquals(exception.getMessage.substring(0, err.length), err)
-    }
-  }
-
-  Seq(
-    // regular or low-surrogate or char codes
+    // regular or half-surrogate char codes
+    55297,
     56375
   ).foreach { ch =>
-    test(s"#3690 non-high-surrogate with char code $ch") {
+    test(s"#3690 any character with char code $ch") {
       tokenize(
         s"""|"${ch.toChar}"
             |""".stripMargin,

